### ISSUE TITLE
Go back/forward in browser history with mouse buttons

### ISSUE THREAD

**Issue by luixxiul** [brave/browser-laptop#3101]
From support:

> I usually use two extra buttons on my mouse to go back and forth in browser history.  It works natively on any other browser on my machine (web developer. got plenty of browsers). Please make it work.

On Arch linux on notebook.

**Comment by bsclifton** [comment]
marked as keyboard-shortcuts, even though it's technically a mouse button :stuck_out_tongue:

**Comment by Aeyoun** [comment]
Confirmed on macOS 10.12 and Fedora Linux. Brave 0.11.6.

**Comment by ilian** [comment]
Related issues: #938 and #1807 
Same problem here. Tested with brave 0.12.1 on Fedora 23, but works fine on a Windows machine.

**Comment by alexwykoff** [comment]
+1 from support : https://linkbubble.zendesk.com/agent/tickets/6334

**Comment by Haegin** [comment]
Confirmed on Brave on Mac OS X El Capital

OS X: 10.11.6 (15G1004)

Brave: 0.12.8
Electron: 1.4.23
libchromiumcontent: 53.0.2785.143
V8: 5.3.332.47
Node.js: 6.5.0
Update Channel: dev

**Comment by Liunkae** [comment]
+1 from #7127.

**Comment by defaye** [comment]
Clicking "back" with my mouse every 2 minutes after remembering this pitfall, and constantly reminding myself how much of a pain in the arse it is not to have this feature.

**Comment by bsclifton** [comment]
+1 from user @Raindeer44 on Linux Mint 18.1 (reported in https://github.com/brave/browser-laptop/issues/1369)

**Comment by sd86** [comment]
+1. Need this to make the browser usable for day-to-day work.

**Comment by ftcmnc** [comment]
+1

**Comment by scmart** [comment]
I really want to use the browser, 0 chance I can do that without forward/back working properly.  This is a deal breaker.

**Comment by drhandwerk** [comment]
Confirmed on Ubuntu 16.04 
Brave 0.16.9

This makes the browser unusable for me. Will have to go back to Firefox with plugins.

**Comment by wcat13** [comment]
Really no fix for this? Yikes, was loving Brave but cannot surf without muh back button.

**Comment by synlic** [comment]
Same issue here. Using freshly installed Archlinux.

**Comment by lukemulks** [comment]
+1 issue reference from basicattentiontoken.rocket.chat contributor. 
https://basicattentiontoken.rocket.chat/channel/brave-browser?msg=ZHpNmonumxDETFdmN

**Comment by avagong** [comment]
Please support this feature. I'm on CentOS 7 with a Logitech M500. The forward/back buttons work on every browser I've tried except Brave.

**Comment by wulfmeister** [comment]
😢

**Comment by wihodges** [comment]
Yea seems super strange that this can't be fixed. 

If you have Logitech mouse [Im using MX] you can go into Logitech Options > Add specific application rule for Brave > Keystroke assignment. Then assign fwd/back buttons in OSX: "command [" for back and "command ]" for forward. Now I can browse just fine.

**Comment by cstyan** [comment]
@bsclifton Any update here? I'm running debian kernel 4.11, steelseries rival mouse, brave 0.18.36. 

Mouse back and forward (xev mouse button 8 and 9) do not work, however they work as expected in firefox and chrome.

**Comment by bsclifton** [comment]
cc: @zentagonist for triage

**Comment by bsclifton** [comment]
+1 from @agilob via https://github.com/brave/browser-laptop/issues/11544

**Comment by TomFaulkner** [comment]
This is still a problem on Linux (Ubuntu 17.04). It's frustrating for sure.

**Comment by srirambv** [comment]
+1 from community https://community.brave.com/t/support-mouse-back-forward-buttons-on-linux/10480/

**Comment by realAkhmed** [comment]
Confirming on Ubuntu 16.04: Brave 0.19.95-1 + mouse Logitech MX510

This is a really major issue. We may be losing a lot of the target demographics (privacy-minded Linux users).

**Comment by realAkhmed** [comment]
Confirming on Kubuntu 16.04: Brave 0.19.95-1 + mouse Logitech MX510. `xev` shows these are buttons 8 and 9.

Found a work-around by mapping buttons 8 and 9 to keyboard Alt-Left and Alt-Right as described here: http://marian.schedenig.name/2012/06/07/mapping-kde-actions-to-extra-mouse-buttons/

More precisely. First do:
```
$ sudo apt-get install xbindkeys xautomation
$ xbindkeys --defaults > ~/.xbindkeysrc
```
then add the following two lines to `.xbindkeysrc`:
```

# Bind "forward" mouse button to Alt - Right
"xte 'keydown Alt_L' 'key Right' 'keyup Alt_L'"
  b:9

# Bind "back" mouse button to Alt - Left
"xte 'keydown Alt_L' 'key Left' 'keyup Alt_L'"
  b:8
```
Then, restart `xbindkeys`
```
$ killall xbindkeys
$ xbindkeys
```

**Comment by quesada** [comment]
This is really irritating. P3? well, no backspace, no mouse button makes the browser useless for me. Two finger shortcut to do an action that I do 100s times a day? No thank you. You just lost a user.

**Comment by mbootsman** [comment]
It's amazing how something simple like this has not been fixed yet. Probably no priority (= P3) because of the low usage stats for Linux users.

**Comment by TomFaulkner** [comment]
It is disappointing that this isn't seeing any love, it's a basic usability feature. I'm on Waterfox for now because of this.

**Comment by KoHoSo** [comment]
As I recall, a similar issue in the early public development of Vivaldi was addressed fairly quickly.  With both that browser and Brave being based on the same Blink engine, I cannot understand why 15 months have passed since the first post with no fix.

Above, somebody guessed this is a low priority for Brave because Linux usage is so low.  Considering how Linux users long for good alternatives for _all_ types of software, who knows how many try Brave and immediately gave up without comment because of this very problem.

What I _do_ know is this.  I give Brave a good try on every major update and this deal-breaker issue pushes me away every time...and not just on my Linux machine but also on my Win7 one at work because, if one is unusable, I don't want to bother with it on the other.  Please fix this.

**Comment by djoreilly** [comment]
Please fix. I have tried many browsers on Linux over the years and I can't remember this not working.

**Comment by TomFaulkner** [comment]
Based on this page: https://github.com/brave/browser-laptop/projects, and the issue being discussed with the mouse back/forward buttons being listed in the "contributor backlog," which has the description: "These are issues we are unlikely to fix internally, but we would love for you to pick up the cause and share your talents with the team!" it seems there isn't a concern for the mouse back/forward buttons under Linux, and it won't be fixed by the dev team, it will only be fixed if a user decides to fix it.

I've switched away from using Brave and use Watefox primarily now, been keeping Brave around hoping it gets better, but this doesn't seem promising to me.

**Comment by TomFaulkner** [comment]
I spent some time looking into the issue today. Electron doesn't seem to provide for a way to read mouse buttons outside of the normal three buttons, so adding this doesn't seem trivial.

**Comment by Alino** [comment]
@TomFaulkner Brave is not using Electron but they have forked it and named it [Muon](https://github.com/brave/muon)
So technically we are able to add this mouse buttons capability into Muon.

**Comment by zacharyparrish** [comment]
+1 for Linux. Running Ubuntu Gnome 17.04

**Comment by TomFaulkner** [comment]
@Alino, you are right, sorry I wasn't clear on it being Muon rather than Electron.

**Comment by srirambv** [comment]
+1 from community 
https://community.brave.com/t/forward-and-backward-buttons-on-mouse-and-hide-the-menu-bar-by-default-do-not-work-in-brave-on-ubuntu-17-10/13159/3

**Comment by Sector14** [comment]
Many issues in a browser I can put up with, but this is such a core usability issue that despite liking the speed and general goal of Brave I've had to uninstall it. 

If a core usability issue for a segment of your user base is ignored for over a year (and apparently with no intention to ever fix), how many other issues including ones that impact me or people I might one day have recommended the browser to, will equally be ignored?

**Comment by Alino** [comment]
it's all about priorities. There is so much work that needs to be done, and this issue has not a huge impact. Most people won't use this feature. Only some part of the users are using an actual desktop with a mouse, and from this group you can create even smaller group that has such a mouse which offers other buttons than classic left & right click & scroll wheel, and from this group which has such a mouse only some of them are actually using these buttons to navigate in a browser.
This is a small selection of users.

**Comment by JReinhold** [comment]
@Alino that is somewhat true, although that can be said about almost any feature. Unless you have some real numbers, you are basically working off assumptions, which could be as good as anyone's. 
It could easily be:

- Brave users: 100%
- Users with a mouse: 80%
- Users with mouse that has back/forward buttons: 70%
- Users that uses back/forward buttons to navigate in browsers: 50%

And now it's suddenly a feature that affects 50% of the user base. From my experience that is actually quite low, as 75% of people I live and work with use this feature daily.

**Comment by Alino** [comment]
@JReinhold 
Laptops sales did overtake desktop sales already in 2008 and since then, we went even more mobile. Majority of laptop people are using trackpad/touchpad not a mouse at all.
So therefore I would say there is much less brave users with a mouse device than 80%. I would say it could be something near 30%.
- Users with mouse that has back/forward buttons: 20%
- Users that uses back/forward buttons to navigate in browsers: 2 - 8%

now we did end up with a small user base, that is impacted by this. There are currently opened issues that impact all or nearly all users. These shall have higher priorities.

**Comment by LaurenWags** [comment]
+1 from twitter: https://twitter.com/meganewsx/status/953715223365435392

**Comment by jonathansampson** [comment]
+1 from Twitter: https://twitter.com/CelmorS/status/955264039625555968

![image](https://user-images.githubusercontent.com/815158/35202668-994ff1b2-fee9-11e7-98f1-1a7cd9b83cf2.png)

**Comment by jolars** [comment]
This turned out to be a dealbreaker for me as well.

**Comment by defaye** [comment]
@Alino and most people that will download Brave are surely going to be power-users to begin with. If you aren't willing to cater for your seed audience, you might as well dump your project IMO. There are far more people with these kind of mice than you think. Far more that are going to be using some relatively unknown alternative web-browser.

**Comment by Alino** [comment]
@defaye you are familiar with javascript right, why don't you contribute if this issue affects you? This project is open source, and if any obstacle comes into your way while you are using it, you can remove that obstacle.

**Comment by defaye** [comment]
@Alino yeah that's very true. I don't wanna be a scrooge on this issue..

**Comment by zroulston** [comment]
+1 for Linux. xorg/gnome - Fedora release 27 (Twenty Seven)

**Comment by bsclifton** [comment]
I don't think this would be terribly hard to implement, if someone was open to grabbing the issue (one of us here can help answer questions)

@jolars @defaye @zroulston what type of mouse do you have? I personally use a two button + scroll-wheel, so I'm unable to try this out... but my mouse is on the way out. Maybe I'll get a new mouse with buttons

**Comment by jolars** [comment]
@bsclifton I have a Razer Deathadder Chroma. Unfortunately I do not have the know-how to provide much help, other than to test it out, but I would love to see it implemented.

**Comment by zroulston** [comment]
@bsclifton I'm using a Logitech G500.

**Comment by defaye** [comment]
@bsclifton I've looked into it and there aren't any events fired for mouse buttons beyond the main 3 (left, centre, right). So I'm not entirely sure how you'd approach this from a JS angle. I figured this may be the main reason this issue has been set aside

**Comment by TomFaulkner** [comment]
@defaye That is what I found when I looked into it a few weeks ago as well. The Muon / Electron documentation and what I found elsewhere seem to indicate that Javascript doesn't handle those other mouse buttons. So, it might be a considerably more difficult problem.

Someone mentioned earlier that Vivaldi similarly didn't support mouse back/forward for some time, but they fixed it. I haven't looked at their source, but perhaps the solution might be similar.

Vivaldi source: https://github.com/ric2b/Vivaldi-browser

**Comment by zroulston** [comment]
For what it's worth, the back button works on macOS 10.13.3.

**Comment by vinicentus** [comment]
The buttons work for me on Windows x64

**Comment by vinicentus** [comment]
TL;DR, but the mouse buttons work for me on Windows x64

**Comment by derN3rd** [comment]
Mouse back button does not work on Ubuntu 16.04 with Brave 0.20.42

**Comment by TomFaulkner** [comment]
With some of the feedback here, it seems the issue name should be changed to refer to the mouse back/forward not working in Linux, rather than not working in general.

@luixxiul, can you change the title?

**Comment by kentonv** [comment]
Showstopper for me. :/

**Comment by jlenz314** [comment]
I was completely sold on Brave, had everything ported over and was loving it...until this. Back to Chrome until this gets fixed.

**Comment by bsclifton** [comment]
+1 from @AtjonTV via https://github.com/brave/browser-laptop/issues/13880
> I am using the [ET D-09](https://www.aliexpress.com/item/ET-Brand-D-09-2-4Ghz-Wireless-Mouse-1600-CPI-6-Buttons-Computer-Mouse-For-Laptops/2017090370.html), its a mouse with a integrated back and forth button. When I try to use either of them in Brave its not working. I used to use them in Vivaldi to make site navigation quicker. This is a feature I would love to see in Brave.

**Comment by TheDevKing** [comment]
It looks like [this](https://cs.chromium.org/chromium/src/chrome/browser/ui/views/frame/browser_command_handler_linux.cc?sq=package:chromium&l=31) is where this specific issue is being addressed in Chromium. Unfortunately I don't know how to go about bringing this into Muon.

**Comment by Liunkae** [comment]
@TheDevKing If this problem is fixed in Chromium, the [new version of Brave](/brave/brave-browser) may get the fix for free.

**Comment by TheDevKing** [comment]
After building the [new version](https://github.com/brave/brave-browser) and testing in Mint 18.3 I can confirm that the back and forward buttons _do_ work as expected (at least for me). Just thought I'd follow up with an actual confirmation of this issue being fixed (again, just going by my single test).

**Comment by naisanza** [comment]
Tested on Ubuntu 18.04, and mouse backwards and forward buttons do not work

**Comment by AtjonTV** [comment]
Buttons working fine on Elementary OS 0.4.1 (Ubuntu 16.04) and the latest version of Brave.

**Comment by bsclifton** [comment]
I'm curious if recent updates (ex: Chromium 67) have fixed this. Care to give it a retry?
@avagong @cstyan @da2x @defaye @djoreilly @drhandwerk @ftcmnc @Haegin @ilian @jlenz314 @jolars @kentonv @scmart @synlic @TomFaulkner @zacharyparrish

**Comment by TheDevKing** [comment]
Works for me on Linux Mint

```
OS: Mint 18.3 sylvia
Kernel: x86_64 Linux 4.13.0-41-generic
```
```
Brave: 0.22.810
V8: 6.7.288.43
rev: 8f30eeb
Muon: 7.0.6
OS Release: 4.13.0-41-generic
Update Channel: Release
OS Architecture: x64
OS Platform: Linux
Node.js: 7.9.0
Brave Sync: v1.4.2
libchromiumcontent | 67.0.3396.71
```

**Comment by da2x** [comment]
Also works for me, Brave 0.22.810 on Fedora 27 (GNOME on Wayland).

**Comment by KoHoSo** [comment]
The back button on Brave 0.23.16 now works for me on Linux Mint 18.3 MATE 64-bit.  This really makes me happy!  Thanks to all of you that kept up on this issue!

**Comment by mbootsman** [comment]
Back button works on Ubuntu 18.04 with Brave version 0.22.810
